report/src/system_design/kbeacondata.tex

```diff
@@ -0,0 +1,166 @@
+	captionpos=b] 
+public sealed class KBeaconData {
+	private const int ByteCount = 25;
+
+	public Guid Uuid { get; }
+
+	public byte[] CompanyId { get; }
+
+	public ushort Major { get; }
+	public ushort Minor { get; }
+
+	public sbyte TxPower { get; }
+
+	public KBeaconData(byte[] data) 
+		: this(data, BitConverter.IsLittleEndian) { }
+
+	public KBeaconData(byte[] data, bool isLittleEndian) {
+		if (data.Length != ByteCount)
+			throw new ArgumentException($"Number of bytes was {data.Length}, expected {ByteCount}.");
+
+		if (isLittleEndian) {
+			CompanyId = data.Take(2).ToArray();
+			Uuid = new Guid(
+				BitConverter.ToInt32(data.Skip(4).Take(4).ToArray()),
+				BitConverter.ToInt16(data.Skip(8).Take(2).ToArray()),
+				BitConverter.ToInt16(data.Skip(10).Take(2).ToArray()),
+				data.Skip(12).Take(8).ToArray()
+			);
+			Major = BitConverter.ToUInt16(data.Skip(20).Take(2).ToArray());
+			Minor = BitConverter.ToUInt16(data.Skip(22).Take(2).ToArray());
+		} else {
+			CompanyId = data.Take(2).ToArray();
+			Uuid = new Guid(
+				BitConverter.ToInt32(data.Skip(4).Take(4).Reverse().ToArray()),
+				BitConverter.ToInt16(data.Skip(8).Take(2).Reverse().ToArray()),
+				BitConverter.ToInt16(data.Skip(10).Take(2).Reverse().ToArray()),
+				data.Skip(12).Take(8).ToArray()
+			);
+			Major = BitConverter.ToUInt16(
+				data.Skip(20).Take(2).Reverse().ToArray()
+			);
+			Minor = BitConverter.ToUInt16(
+				data.Skip(22).Take(2).Reverse().ToArray()
+			);
+		}
+		TxPower = (sbyte)data[^1];
+	}
+}
+\end{lstlisting}
+\end{figure}
+
+The notion of endianness is crucial while interpreting the byte sequence.
+It refers to the byte order in which the integers are stored in the computer memory. In the context of the KBeaconData class, the constructor accepts a boolean parameter that indicates whether the byte order of the data is little-endian or not. If the byte order is indeed little-endian, the sequence is read as-is. Otherwise, the sequence is reversed before interpreting, thereby maintaining the accuracy of the parsed data.
+
+In both little and big-endian scenarios, the first two bytes are extracted as the Company ID.
+However, the remaining data is parsed differently depending on the endianness. If the byte order is little-endian, the bytes are sequentially processed. The UUID is constructed from the next 16 bytes, with the first four bytes converted to an integer, the next four bytes to two short integers, and the last eight bytes remain as a byte array. The Major and Minor values are derived from the next two bytes each, interpreted as unsigned short integers. In case of big-endian, the sequence is reversed before parsing the UUID, Major, and Minor values.
+
+The last byte, regardless of the endianness, is directly converted to a signed byte to represent the TxPower.
+
+
+The KBeaconData class thereby provides an effecient way to parse, encapsulate, and interpret the data from beacon advertisement packets.
+
```

Comment:
Maybe consider a transition or describe the order of the sections in the system design intro